### ISSUE TITLE
[Phase 4] Transfer Learning for Multi-Equipment

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/predictive-maintenance-survival#5]
## Summary

Implement transfer learning to share learned representations across different equipment types (C-MAPSS subsets FD001-FD004), reducing data requirements for new machines.

**Parent Issue:** #1
**Depends on:** #4 (Phase 3: RUL Estimation)

## Objectives

- Learn shared representations from source equipment
- Fine-tune on target equipment with limited data
- Detect and prevent negative transfer
- Compare transfer vs from-scratch performance

## Key Concepts

**Domain Shift**: FD001-FD004 have different operating conditions and fault modes. Direct application may fail due to distribution mismatch.

**Negative Transfer**: When source knowledge hurts target performance. Must detect and fallback to from-scratch training.

## Files to Create

| File | Description |
|------|-------------|
| `src/models/transfer_learning.py` | Multi-equipment transfer model |
| `src/models/domain_adaptation.py` | Distribution matching techniques |
| `configs/equipment_config.yaml` | Equipment type definitions |
| `tests/test_transfer_learning.py` | Transfer learning tests |

### Transfer Model
- [ ] Shared feature encoder across equipment types
- [ ] Equipment-specific prediction heads
- [ ] Pretraining on source domain(s)
- [ ] Fine-tuning API with encoder freezing option

### Domain Adaptation
- [ ] Compute domain similarity metrics (MMD, correlation)
- [ ] Weighted combination of source models
- [ ] Operating condition normalization

### Negative Transfer Detection
- [ ] Monitor validation performance during fine-tuning
- [ ] Early stopping if performance degrades
- [ ] Fallback to from-scratch training
- [ ] Warning system for users

### Experiments
- [ ] FD001 → FD002 transfer (same fault, different conditions)
- [ ] FD001 → FD003 transfer (same conditions, different faults)
- [ ] Multi-source transfer (combine FD001+FD002 → FD003)

## Code Skeleton

📄 **File:** `src/models/transfer_learning.py`

```python
"""Transfer learning for multi-equipment survival models."""

import numpy as np
from typing import Optional, Tuple
from sklearn.preprocessing import StandardScaler


class TransferSurvivalModel:
    """Survival model with transfer learning capabilities."""

    def __init__(self, base_model_class, **model_kwargs):
        """
        Args:
            base_model_class: Class of survival model to use
            **model_kwargs: Arguments passed to base model
        """
        self.base_model_class = base_model_class
        self.model_kwargs = model_kwargs

        self.source_model = None
        self.target_model = None
        self.source_scaler = StandardScaler()
        self.target_scaler = StandardScaler()
        self.is_pretrained = False

    def pretrain(
        self,
        X_source: np.ndarray,
        duration_source: np.ndarray,
        event_source: np.ndarray
    ):
        """Pretrain on source domain data.

        Args:
            X_source: Source domain features
            duration_source: Time to event/censoring
            event_source: Event indicators
        """
        # Fit scaler on source
        X_scaled = self.source_scaler.fit_transform(X_source)

        # Train source model
        self.source_model = self.base_model_class(**self.model_kwargs)
        self.source_model.fit(X_scaled, duration_source, event_source)
        self.is_pretrained = True

        return self

    def finetune(
        self,
        X_target: np.ndarray,
        duration_target: np.ndarray,
        event_target: np.ndarray,
        validation_data: Optional[Tuple] = None,
        use_source_scaler: bool = True
    ) -> dict:
        """Fine-tune on target domain.

        Args:
            X_target: Target domain features
            duration_target: Target time to event
            event_target: Target event indicators
            validation_data: Optional (X_val, dur_val, event_val) for monitoring
            use_source_scaler: Whether to use source domain normalization

        Returns:
            Dict with 'success', 'source_score', 'target_score', 'improved'
        """
        if not self.is_pretrained:
            raise RuntimeError("Must call pretrain() before finetune()")

        # Scale target data
        if use_source_scaler:
            X_scaled = self.source_scaler.transform(X_target)
        else:
            X_scaled = self.target_scaler.fit_transform(X_target)

        # Evaluate source model on target (before fine-tuning)
        source_score = self.source_model.score(
            X_scaled, duration_target, event_target
        )

        # Train fresh model on target for comparison
        from_scratch = self.base_model_class(**self.model_kwargs)
        from_scratch.fit(X_scaled, duration_target, event_target)
        scratch_score = from_scratch.score(
            X_scaled, duration_target, event_target
        )

        # Fine-tune: initialize from source, train on target
        # For tree models, we just train on target (no weight transfer)
        self.target_model = self.base_model_class(**self.model_kwargs)
        self.target_model.fit(X_scaled, duration_target, event_target)
        target_score = self.target_model.score(
            X_scaled, duration_target, event_target
        )

        # Detect negative transfer
        improved = target_score > scratch_score

        return {
            'success': True,
            'source_score_on_target': source_score,
            'from_scratch_score': scratch_score,
            'finetuned_score': target_score,
            'improved_over_scratch': improved,
            'transfer_gain': target_score - scratch_score
        }

    def predict_survival_function(self, X: np.ndarray, use_target: bool = True):
        """Predict using fine-tuned or source model."""
        model = self.target_model if use_target else self.source_model
        if model is None:
            raise RuntimeError("Model not trained")

        scaler = self.target_scaler if use_target else self.source_scaler
        X_scaled = scaler.transform(X)
        return model.predict_survival_function(X_scaled)


def compute_domain_similarity(
    X_source: np.ndarray,
    X_target: np.ndarray
) -> dict:
    """Compute similarity metrics between source and target domains.

    Returns:
        Dict with 'mmd' (Maximum Mean Discrepancy), 'feature_correlations'
    """
    from scipy.spatial.distance import cdist

    # Simple MMD approximation using mean embeddings
    source_mean = X_source.mean(axis=0)
    target_mean = X_target.mean(axis=0)
    mmd = np.linalg.norm(source_mean - target_mean)

    # Per-feature correlation between domains
    correlations = []
    for j in range(X_source.shape[1]):
        src_vals = X_source[:, j]
        tgt_vals = X_target[:, j]

        # Compare distributions via overlap
        src_min, src_max = src_vals.min(), src_vals.max()
        tgt_min, tgt_max = tgt_vals.min(), tgt_vals.max()

        overlap = max(0, min(src_max, tgt_max) - max(src_min, tgt_min))
        range_union = max(src_max, tgt_max) - min(src_min, tgt_min)
        correlations.append(overlap / range_union if range_union > 0 else 0)

    return {
        'mmd': mmd,
        'feature_overlap': np.mean(correlations),
        'per_feature_overlap': correlations
    }
```

## Definition of Done

- [ ] Transfer model outperforms from-scratch with < 50% target data
- [ ] Fine-tuning works with as few as 10 target engines
- [ ] Negative transfer detection triggers warning
- [ ] Domain similarity metrics computed correctly
- [ ] All tests pass

## Experiments to Run

| Experiment | Source | Target | Expected Outcome |
|------------|--------|--------|------------------|
| Same fault, diff conditions | FD001 | FD002 | Moderate transfer |
| Same conditions, diff faults | FD001 | FD003 | Limited transfer |
| Multi-source | FD001+FD002 | FD004 | Best transfer |

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Negative transfer | 🔴 | Detect and warn user |
| Distribution shift too large | 🟡 | Domain adaptation techniques |
| Insufficient target data | 🟡 | More aggressive regularization |
